src/concatenate.jl

```diff
@@ -0,0 +1,163 @@
+"""
+    module Concatenate
+
+Alternative implementation for `Base.cat` through [`concatenate(!)`](@ref).
+
+This is mostly a copy of the Base implementation, with the main difference being
+that the destination is chosen based on all inputs instead of just the first.
+
+Additionally, we have an intermediate representation in terms of a Concatenated object,
+reminiscent of how Broadcast works.
+
+The various entry points for specializing behavior are:
+
+* Destination selection can be achieved through
+
+    Base.similar(cat::Concatenated{Interface}, ::Type{T}, axes) where {Interface}
+
+* Implementation for moving one or more arguments into the destionation through
+
+    copy_offset!(dest, shape, catdims, offsets, args...)
+    copy_offset1!(dest, shape, catdims, offsets, x)
+
+* Custom implementations:
+
+    Base.copy(cat::Concatenated{Interface}) # custom implementation of concatenate
+    Base.copyto!(dest, cat::Concatenated{Interface}) # custom implementation of concatenate! based on interface
+    Base.copyto!(dest, cat::Concatenated{Nothing}) # custom implementation of concatenate! based on typeof(dest)
+"""
+module Concatenate
+
+using Compat: @compat
+
+export concatenate, concatenate!
+@compat public Concatenated, cat_offset!, cat_offset1!, copy_or_fill!
```

Comment:
I'm a bit hesitant about making functions like `cat_offset!`, `cat_offset1!`, and `copy_or_fill!` public. I see the point is that they can act as lower level specialization points, but for now can we hold off on that and see how it gets used in practice?

Reply:
We can, but in that case there really is not that much reason to have that part of the implementation in the first place. If we don't want to expose this, I can simply couple back to `Base.cat_offset!` in the `copyto!` function that implements this, and not even have to copy these Base implementations. I'm definitely also okay with that, but then I would much rather just delete these lower-level functions altogether.

Reply:
That sounds good to delete those low-level functions, I think biasing towards keeping things simpler right now is better and we can always add those things back if needed.